src/vision/status_quo/alan_hates_writing_a_stream.md

```diff
@@ -0,0 +1,194 @@
+
+He thinks about his stream's states, and settles on the following structure:
+
+```rust
+struct SigningFile {
+    state: State,
+    file: ReaderStream,
+    sig: Signature,
+}
+
+enum State {
+    File,
+    Sign,
+}
+```
+
+Now he tries to write the `poll_next` method, checking readiness of individual steps and proceeding to the next state, while grumbling away the weird `Pin` noise:
+
+```rust
+todo!("is this worth showing in the story, or just noise?")
+```
+
+It turns out it was more complicated than Alan thought (the author made this same mistake). The `digest` method of `Signature` is `async`, _and_ it consumes the signature, so the state machine needs to be adjusted. The signature needs to be able to be moved out, and it needs to be able to store a future from an `async fn`. Trying to figure out how to represent that in the type system was difficult (expand?), but Alan eventually learns to just store a `Pin<Box<dyn Future>>`, wondering if the `Pin` there is important.
+
+```rust
+struct SigningFile {
+    state: State,
+}
+
+enum State {
+    File(ReaderStream, Signature),
+    Signing(Pin<Box<dyn Future<Item = Vec<u8>>>>),
+    Done,
+}
+```
+
+(show how gross this codes gets?)
+
+Oh well, at least it _works_, right?
+
+#### Wakers
+
+So far, Alan hasn't paid too much attention to `Context` and `Poll`. It's been fine to simply pass them along untouched. There's a confusing bug in his state machine. Let's look more closely:
+
+```rust
+// zooming in!
+match ready!(file.poll_next(cx)) {
+    Some(val) => {
+        me.sig.as_mut().unwrap().push(val.len()));
+        return Poll::Ready(Some(val));
+    },
+    None => {
+        me.state = State::Sign;
+        // oops!
+        return Poll::Pending;
+    }
+}
+```
+
+In one of the branches, the state is changed, and `Poll::Pending` is returned. Alan assumes that the task will be polled again with the new state. But, since the file was done (and has returned `Poll::Ready`), there was actually no waker registered to wake the task again. So his stream just hangs forever.
+
+The compiler doesn't help at all, and he re-reads his code multiple times, but because of this easy-to-misunderstand logic error, Alan eventually has to ask for help in a chat room. After a half hour of explaining all sorts of details, a kind person points out he either needs to register a waker, or perhaps use a loop.
+
+(Show the solution? in far too many cases (likely this one), the code gets turned into a weird loop, or the user must learn about wakers).
+
+### Gives Up
+
+A little later, Alan needs to add some response body transforming to some routes, to add some app-specific framing. Upon realizing he needs to implement another `Stream` in a generic fashion, he instead closes the editor and complains on Twitter.
+
+
+
+## 🤔 Frequently Asked Questions
+
+
+* **What are the morals of the story?**
+    * Writing an async `Stream` is drastically different than writing an `async fn`.
```

Comment:
I'm worried that this story could be shrugged off as "landing generators will fix this". Can we also make a note that the primitives that Rust's async story uses are not super straight-forward to compose?

Reply:
It's probable that generators are indeed the "shiny future", but there's two more bullet points about `Pin` and `Waker`. What else about "primitives" did you have in mind as morals?

Reply:
I was thinking the same thing, I expect generators *are* a big part of the shiny future. I guess that one thing which might help would be to write a second story, or expand this story, to talk about implementation a custom future -- or perhaps add a FAQ like "Besides streams, where else might users encounter pin, context, and friends?"

Reply:
I would rather think about a shiny future which doesn't involve more special types but less. The whole thing could also be avoided by 

```rust
trait Stream {
    type Item;
    async fn next(&mut self) -> Option<Self::Item>;
}
```

We are all aware that this one also has challenges. But it applies for all composed async operations.

But since the goal of this doc is to showcase the challenge and not solutions, it seems good as is.